File: selenium-tests/src/main/java/com/timesheet/pages/ClientsPage.java
package com.timesheet.pages;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;

import java.util.List;

public class ClientsPage extends BasePage {

    private final By pageTitle = By.xpath("//h4[contains(text(), 'Clients')]");
    private final By addClientButton = By.xpath("//button[contains(., 'Add Client')]");
    private final By clearAllButton = By.xpath("//button[contains(., 'Clear All')]");
    private final By clientsTable = By.cssSelector("table");
    private final By clientRows = By.cssSelector("table tbody tr");
    private final By noClientsMessage = By.xpath("//*[contains(text(), 'No clients found')]");
    private final By loadingSpinner = By.cssSelector(".MuiCircularProgress-root");
    private final By errorAlert = By.cssSelector(".MuiAlert-standardError");

    private final By dialogTitle = By.cssSelector(".MuiDialogTitle-root");
    private final By clientNameInput = By.cssSelector("input[label='Client Name'], .MuiDialog-root input:first-of-type");
    private final By departmentInput = By.cssSelector("input[label='Department']");
    private final By emailInput = By.cssSelector("input[type='email']");
    private final By descriptionInput = By.cssSelector("textarea");
    private final By cancelButton = By.xpath("//button[contains(text(), 'Cancel')]");
    private final By createButton = By.xpath("//button[contains(text(), 'Create')]");
    private final By updateButton = By.xpath("//button[contains(text(), 'Update')]");
    private final By dialogLoadingSpinner = By.cssSelector(".MuiDialog-root .MuiCircularProgress-root");

    private final By editButtons = By.cssSelector("button[color='primary'] svg[data-testid='EditIcon'], button svg[data-testid='EditIcon']");
    private final By deleteButtons = By.cssSelector("button[color='error'] svg[data-testid='DeleteIcon'], button svg[data-testid='DeleteIcon']");

    public ClientsPage(WebDriver driver) {
        super(driver);
    }

    public boolean isClientsPageDisplayed() {
        try {
            waitUtils.waitForElementVisible(pageTitle);
            return isElementDisplayed(pageTitle);
        } catch (Exception e) {
            return false;
        }
    }

    public void waitForPageLoad() {
        waitUtils.waitForElementInvisible(loadingSpinner);
        waitUtils.waitForElementVisible(pageTitle);
        waitUtils.sleep(500);
    }

    public void clickAddClientButton() {
        click(addClientButton);
        waitUtils.waitForElementVisible(dialogTitle);
    }

    public boolean isClearAllButtonDisplayed() {
        return isElementDisplayed(clearAllButton);
    }

    public void clickClearAllButton() {
        click(clearAllButton);
    }

    public void confirmClearAll() {
        acceptAlert();
        waitUtils.sleep(1000);
    }

    public void clearAllClients() {
        if (isClearAllButtonDisplayed()) {
            clickClearAllButton();
            confirmClearAll();
            waitForPageLoad();
        }
    }

    public int getClientCount() {
        if (isElementDisplayed(noClientsMessage)) {
            return 0;
        }
        List<WebElement> rows = getElements(clientRows);
        if (rows.size() == 1 && rows.get(0).getText().contains("No clients found")) {
            return 0;
        }
        return rows.size();
    }

    public boolean hasNoClients() {
        return isElementDisplayed(noClientsMessage) || getClientCount() == 0;
    }

    public void enterClientName(String name) {
        List<WebElement> inputs = driver.findElements(By.cssSelector(".MuiDialog-root input"));
        if (!inputs.isEmpty()) {
            inputs.get(0).clear();
            inputs.get(0).sendKeys(name);
        }
    }

    public void enterDepartment(String department) {
        List<WebElement> inputs = driver.findElements(By.cssSelector(".MuiDialog-root input"));
        if (inputs.size() > 1) {
            inputs.get(1).clear();
            inputs.get(1).sendKeys(department);
        }
    }

    public void enterEmail(String email) {
        List<WebElement> inputs = driver.findElements(By.cssSelector(".MuiDialog-root input"));
        if (inputs.size() > 2) {
            inputs.get(2).clear();
            inputs.get(2).sendKeys(email);
        }
    }

    public void enterDescription(String description) {
        WebElement textarea = driver.findElement(descriptionInput);
        textarea.clear();
        textarea.sendKeys(description);
    }

    public void clickCreateButton() {
        click(createButton);
        waitUtils.waitForElementInvisible(dialogLoadingSpinner);
        waitUtils.sleep(500);
    }

    public void clickUpdateButton() {
        click(updateButton);
        waitUtils.waitForElementInvisible(dialogLoadingSpinner);
        waitUtils.sleep(500);
    }

    public void clickCancelButton() {
        click(cancelButton);
        waitUtils.sleep(300);
    }

    public void createClient(String name, String department, String email, String description) {
        clickAddClientButton();
        enterClientName(name);
        if (department != null && !department.isEmpty()) {
            enterDepartment(department);
        }
        if (email != null && !email.isEmpty()) {
            enterEmail(email);
        }
        if (description != null && !description.isEmpty()) {
            enterDescription(description);
        }
        clickCreateButton();
        waitForPageLoad();
    }

    public void createClient(String name) {
        createClient(name, null, null, null);
    }

    public boolean isClientInTable(String clientName) {
        List<WebElement> rows = getElements(clientRows);
        for (WebElement row : rows) {
            if (row.getText().contains(clientName)) {
                return true;
            }
        }
        return false;
    }

    public void editClient(String clientName) {
        List<WebElement> rows = getElements(clientRows);
        for (int i = 0; i < rows.size(); i++) {
            if (rows.get(i).getText().contains(clientName)) {
                List<WebElement> editBtns = driver.findElements(By.cssSelector("table tbody tr:nth-child(" + (i + 1) + ") button"));
                for (WebElement btn : editBtns) {
                    if (btn.getAttribute("innerHTML").contains("EditIcon") || 
                        btn.findElements(By.cssSelector("svg[data-testid='EditIcon']")).size() > 0) {
                        btn.click();
                        waitUtils.waitForElementVisible(dialogTitle);
                        return;
                    }
                }
                if (!editBtns.isEmpty()) {
                    editBtns.get(0).click();
                    waitUtils.waitForElementVisible(dialogTitle);
                    return;
                }
            }
        }
    }

    public void deleteClient(String clientName) {
        List<WebElement> rows = getElements(clientRows);
        for (int i = 0; i < rows.size(); i++) {
            if (rows.get(i).getText().contains(clientName)) {
                List<WebElement> deleteBtns = driver.findElements(By.cssSelector("table tbody tr:nth-child(" + (i + 1) + ") button"));
                for (WebElement btn : deleteBtns) {
                    if (btn.getAttribute("innerHTML").contains("DeleteIcon") ||
                        btn.findElements(By.cssSelector("svg[data-testid='DeleteIcon']")).size() > 0) {
                        btn.click();
                        acceptAlert();
                        waitUtils.sleep(500);
                        return;
                    }
                }
                if (deleteBtns.size() > 1) {
                    deleteBtns.get(1).click();
                    acceptAlert();
                    waitUtils.sleep(500);
                    return;
                }
            }
        }
    }

    public void updateClient(String newName, String newDepartment, String newEmail, String newDescription) {
        if (newName != null && !newName.isEmpty()) {
            enterClientName(newName);
        }
        if (newDepartment != null) {
            enterDepartment(newDepartment);
        }
        if (newEmail != null) {
            enterEmail(newEmail);
        }
        if (newDescription != null) {
            enterDescription(newDescription);
        }
        clickUpdateButton();
        waitForPageLoad();
    }

    public String getErrorAlertText() {
        try {
            return getText(errorAlert);
        } catch (Exception e) {
            return null;
        }
    }

    public boolean isErrorAlertDisplayed() {
        return isElementDisplayed(errorAlert);
    }

    public void closeErrorAlert() {
        try {
            WebElement closeBtn = driver.findElement(By.cssSelector(".MuiAlert-standardError button"));
            closeBtn.click();
        } catch (Exception e) {
        }
    }

    public List<String> getClientNames() {
        List<WebElement> rows = getElements(clientRows);
        return rows.stream()
                .map(row -> {
                    List<WebElement> cells = row.findElements(By.tagName("td"));
                    return cells.isEmpty() ? "" : cells.get(0).getText();
                })
                .filter(name -> !name.isEmpty() && !name.contains("No clients found"))
                .toList();
    }

    public boolean isDialogOpen() {
        return isElementDisplayed(dialogTitle);
    }

    public String getDialogTitle() {
        try {
            return getText(dialogTitle);
        } catch (Exception e) {
            return null;
        }
    }
}

File: e2e-tests/page-objects/ClientsPage.js
import BasePage from './BasePage.js';

class ClientsPage extends BasePage {
    get pageHeader() { return $('h4'); }
    get addClientButton() { return $('//button[contains(text(),"Add Client")]'); }
    get clientsTable() { return $('table'); }
    get dialogTitle() { return $('.MuiDialogTitle-root'); }
    get clientNameInput() { return $('//label[contains(text(),"Client Name")]/following-sibling::div//input'); }
    get departmentInput() { return $('//label[contains(text(),"Department")]/following-sibling::div//input'); }
    get emailInput() { return $('//label[contains(text(),"Email")]/following-sibling::div//input'); }
    get descriptionInput() { return $('//label[contains(text(),"Description")]/following-sibling::div//textarea'); }
    get createButton() { return $('//button[contains(text(),"Create")]'); }
    get updateButton() { return $('//button[contains(text(),"Update")]'); }
    get cancelButton() { return $('//button[contains(text(),"Cancel")]'); }
    get errorAlert() { return $('.MuiAlert-standardError'); }
    get noClientsMessage() { return $('//p[contains(text(),"No clients found")]'); }
    get dialog() { return $('.MuiDialog-root'); }

    async open() {
        await super.open('/clients');
    }

    async getPageHeader() {
        const header = await this.pageHeader;
        await header.waitForDisplayed();
        return await header.getText();
    }

    async isAddClientButtonDisplayed() {
        const button = await this.addClientButton;
        return await button.isDisplayed();
    }

    async clickAddClientButton() {
        const button = await this.addClientButton;
        await button.waitForDisplayed();
        await button.click();
        await browser.pause(500);
    }

    async getDialogTitle() {
        const title = await this.dialogTitle;
        await title.waitForDisplayed();
        return await title.getText();
    }

    async enterClientName(name) {
        const input = await this.clientNameInput;
        await input.waitForDisplayed();
        await input.clearValue();
        if (name && name.trim()) {
            await input.setValue(name);
        }
    }

    async clearClientNameField() {
        const input = await this.clientNameInput;
        await input.clearValue();
        await browser.keys(['Tab']);
    }

    async enterDepartment(department) {
        const input = await this.departmentInput;
        await input.waitForDisplayed();
        await input.clearValue();
        if (department) {
            await input.setValue(department);
        }
    }

    async enterEmail(email) {
        const input = await this.emailInput;
        await input.waitForDisplayed();
        await input.clearValue();
        if (email) {
            await input.setValue(email);
        }
    }

    async enterDescription(description) {
        const input = await this.descriptionInput;
        await input.waitForDisplayed();
        await input.clearValue();
        if (description) {
            await input.setValue(description);
        }
    }

    async clickCreateButton() {
        const button = await this.createButton;
        await button.waitForDisplayed();
        await button.click();
        await browser.pause(500);
    }

    async clickUpdateButton() {
        const button = await this.updateButton;
        await button.waitForDisplayed();
        await button.click();
        await browser.pause(500);
    }

    async clickCancelButton() {
        const button = await this.cancelButton;
        await button.waitForDisplayed();
        await button.click();
        await browser.pause(300);
    }

    async isClientInTable(clientName) {
        try {
            const client = await $(`//td//p[contains(text(),"${clientName}")]`);
            return await client.isDisplayed();
        } catch {
            return false;
        }
    }

    async getClientDepartment(clientName) {
        const row = await $(`//td//p[contains(text(),"${clientName}")]/ancestor::tr`);
        const departmentCell = await row.$('td:nth-child(2)');
        return await departmentCell.getText();
    }

    async clickEditButtonForClient(clientName) {
        const row = await $(`//td//p[contains(text(),"${clientName}")]/ancestor::tr`);
        const editButton = await row.$('button[color="primary"]');
        await editButton.click();
        await browser.pause(500);
    }

    async clickDeleteButtonForClient(clientName) {
        const row = await $(`//td//p[contains(text(),"${clientName}")]/ancestor::tr`);
        const deleteButton = await row.$('button[color="error"]');
        await deleteButton.click();
    }

    async confirmDeletion() {
        await browser.acceptAlert();
        await browser.pause(500);
    }

    async cancelDeletion() {
        await browser.dismissAlert();
    }

    async getErrorMessage() {
        const alert = await this.errorAlert;
        await alert.waitForDisplayed({ timeout: 5000 });
        return await alert.getText();
    }

    async isErrorDisplayed() {
        try {
            const alert = await this.errorAlert;
            return await alert.isDisplayed();
        } catch {
            return false;
        }
    }

    async isNoClientsMessageDisplayed() {
        try {
            const message = await this.noClientsMessage;
            return await message.isDisplayed();
        } catch {
            return false;
        }
    }

    async isDialogDisplayed() {
        try {
            const dialog = await this.dialog;
            return await dialog.isDisplayed();
        } catch {
            return false;
        }
    }

    async createClient(name, department = '', email = '', description = '') {
        await this.clickAddClientButton();
        await this.enterClientName(name);
        if (department) await this.enterDepartment(department);
        if (email) await this.enterEmail(email);
        if (description) await this.enterDescription(description);
        await this.clickCreateButton();
        await browser.pause(500);
    }
}

export default new ClientsPage();
